pkg/util/ip_reservation.go

```diff
@@ -0,0 +1,186 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"fmt"
+	"math"
+	"net"
+	"sync"
+)
+
+const (
+	ipRangeSize = 29
+	byteMax     = 255
+	ipV4Bits    = 32
+)
+
+// IPAllocator struct consists of shared resources that are used to keep track of the /29 IPRanges currently reserved by service instances
+type IPAllocator struct {
+	// pendingIPRanges set maintains the set of  IP ranges that have been reserved by the service instances but pending reservation in the cloud instances
+	// The key is a IP range currently reserved by a service instance e.g(192.168.92.0/29). Value is a bool to implement map as a set
+	pendingIPRanges map[string]bool
+
+	// pendingIPRangesMutex is used to synchronize access to the pendingIPRanges set to prevent data races
+	pendingIPRangesMutex sync.Mutex
+}
+
+// NewIPAllocator is the constructor to initialize the IPAllocator object
+// Argument pendingIPRanges map[string]bool is a set of  IP ranges currently reserved by service instances but pending reservation in the cloud instances
+func NewIPAllocator(pendingIPRanges map[string]bool) *IPAllocator {
+	return &IPAllocator{
+		pendingIPRanges: pendingIPRanges,
+	}
+}
+
+// holdIPRange adds a particular IP range in the pendingIPRanges set
+// Argument ipRange string is an IPV4 range which needs put in pendingIPRanges
+func (ipAllocator *IPAllocator) holdIPRange(ipRange string) {
+	ipAllocator.pendingIPRanges[ipRange] = true
+}
+
+// ReleaseIPRange releases the pending IPRange
+// Argument ipRange string is an IPV4 range which needs to be released
+func (ipAllocator *IPAllocator) ReleaseIPRange(ipRange string) {
+	ipAllocator.pendingIPRangesMutex.Lock()
+	defer ipAllocator.pendingIPRangesMutex.Unlock()
+	delete(ipAllocator.pendingIPRanges, ipRange)
+}
+
+// ValidateCIDR function validates whether a particular cidr is a valid IP range
+// Argument cidr string is is a CIDR range that needs to be validated
+func (ipAllocator *IPAllocator) ValidateCIDR(cidr string) bool {
+	ip, ipnet, err := net.ParseCIDR(cidr)
+	if err != nil {
+		return false
```

Comment:
@davidz627 I know that you commented on having this function return a bool instead of an err, but I think given that there are 3 different causes for invalid cidrs, it would be better to return an error that describes each reason.